allennlp/training/deepspeed/checkpointer.py

```diff
@@ -0,0 +1,131 @@
+from typing import Union, Dict, Any, List, Tuple, Optional
+
+import logging
+import os
+import re
+import shutil
+import time
+
+from pathlib import Path
+
+import torch
+
+import allennlp
+from allennlp.nn import util as nn_util
+from allennlp.training import util as training_util, Checkpointer
+
+logger = logging.getLogger(__name__)
+_DeepspeedTrainer = "allennlp.training.deepspeed.trainer.DeepspeedTrainer"
+
+
+class DeepspeedCheckpointer(Checkpointer):
+    # def maybe_save_checkpoint(
+    #     self, 
+    #     trainer: _DeepspeedTrainer, 
+    #     epoch: int, 
+    #     batches_this_epoch: int
+    # ) -> None:
+    #     0/0
+
+    def save_checkpoint(
+        self,
+        epoch: Union[int, str],
+        trainer: _DeepspeedTrainer,
+        is_best_so_far: bool = False,
+        save_model_only=False,
+    ) -> None:
+        if self._serialization_dir is None:
+            return
+
+        with trainer.get_checkpoint_state() as state:
+            model_engine, model_state, training_states = state
+            
+            checkpoint_id = "deepspeed_epoch_{}".format(epoch)
+            model_path = os.path.join(self._serialization_dir, "model_state_epoch_{}".format(epoch))
+            model_engine.save_checkpoint(self._serialization_dir, checkpoint_id)
+
+            # TODO
+            # Model will need a weight file to load; 
+            # not sure if ZeRO stage 2 will mess this up
+            if not os.path.isfile(model_path):
+                    torch.save(model_state, model_path)
```

Comment:
This would be good to know. Have you tried the checkpointing logic? Does it work?

Reply:
The checkpointing works for saving; it's able to go through the training process E2E, doing the checkpointing and so on. I'm just not sure how model-parallel affects this part, if it's saving the entire model state or just the state local to that device. I imagine that this could be validated in a test case.

Reply:
That seems important. I don't want to make a release claiming that this works and then it doesn't in a fairly common use case.